oauthproxy.go

```diff
@@ -578,9 +580,17 @@ func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error)
 	if req.Form.Get("rd") != "" {
 		redirect = req.Form.Get("rd")
 	}
+	// Quirk: On reverse proxies that doesn't have support for
+	// "X-Auth-Request-Redirect" header or dynamic header/query string
+	// manipulation (like Traefik v1 and v2), we can try if the header
+	// X-Forwarded-Host exists or not.
+	if redirect == "" && p.ReverseProxy && isForwardedRequest(req) {
```

Comment:
I think `p.ReverseProxy` should be an argument of `isForwardedRequest`:

Us toggling that we are behind a reverse proxy + seeing the `X-Forwarded-*` headers is the true marker of `isForwardedRequest`.

If we see the `X-Forwarded-*` headers in a request but we aren't behind a reverse proxy, the request isn't really forwarded: the client potentially added those headers themselves.

Reply:
Alrighty!

Reply:
This should also be first in the chain, leaving `X-Auth-Request-Redirect` logic second and overriding `rd` querystring logic last.

Then you don't need the `redirect == ""` bit, makes it more clear that `rd` will take precedence over everything else.

Reply:
Actually yes -- this should definitely be first. Nginx will also send the `X-Forwarded-Host` header, but when used as a K8s ingress, there's an annotation that sets the `X-Auth-Request-Redirect` header to control the redirect (potentially based on Nginx variables, e.g. `$host` or `$escaped_request_uri`)

Reply:
Lastly, we need test cases for the interactions between all 3 paradigms to make sure we don't step on toes. Add some cases to `TestGetRedirect` here: https://github.com/oauth2-proxy/oauth2-proxy/blob/2ceac96677d30ce6b606d1771ba90ca6c0f51c47/oauthproxy_test.go#L1751

Obviously a new case for valid Traefik headers -- but I think also good would be test cases that intermix all types of headers (or with some headers missing) and some `rd` querystring thrown in.